File: ros2_ws/src/perception_pkg/perception_pkg/depth_camera_node.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image, CameraInfo
from std_msgs.msg import Header
from cv_bridge import CvBridge
import numpy as np

try:
    import pyrealsense2 as rs
    REALSENSE_AVAILABLE = True
except ImportError:
    REALSENSE_AVAILABLE = False


class DepthCameraNode(Node):
    def __init__(self):
        super().__init__('depth_camera_node')

        # Parameters
        self.declare_parameter('width', 640)
        self.declare_parameter('height', 480)
        self.declare_parameter('fps', 30)
        self.declare_parameter('enable_depth', True)
        self.declare_parameter('enable_color', True)
        self.declare_parameter('align_depth_to_color', True)
        self.declare_parameter('depth_scale_publish', True)  # /dori/camera/depth_scale
        self.declare_parameter('topics.color_pub', '/dori/camera/color/image_raw')
        self.declare_parameter('topics.depth_pub', '/dori/camera/depth/image_raw')
        self.declare_parameter('topics.depth_colormap_pub', '/dori/camera/depth/image_colormap')
        self.declare_parameter('topics.color_info_pub', '/dori/camera/color/camera_info')
        self.declare_parameter('topics.depth_info_pub', '/dori/camera/depth/camera_info')

        self.width = self.get_parameter('width').value
        self.height = self.get_parameter('height').value
        self.fps = self.get_parameter('fps').value
        self.enable_depth = self.get_parameter('enable_depth').value
        self.enable_color = self.get_parameter('enable_color').value
        self.align_to_color = self.get_parameter('align_depth_to_color').value

        if not REALSENSE_AVAILABLE:
            self.get_logger().error('pyrealsense2 has not been found. Please install the Intel RealSense SDK and pyrealsense2 Python bindings. pip install pyrealsense2')
            return

        # Initialize RealSense pipeline
        self.pipeline = rs.pipeline()
        config = rs.config()

        if self.enable_color:
            config.enable_stream(rs.stream.color, self.width, self.height, rs.format.bgr8, self.fps)
        if self.enable_depth:
            config.enable_stream(rs.stream.depth, self.width, self.height, rs.format.z16, self.fps)

        try:
            profile = self.pipeline.start(config)
        except Exception as e:
            self.get_logger().error(f'Failed to start RealSense pipeline: {e}')
            return

        # Align depth to color if enabled
        self.align = rs.align(rs.stream.color) if self.align_to_color else None

        # depth scale (unit: meter / count)
        depth_sensor = profile.get_device().first_depth_sensor()
        self.depth_scale = depth_sensor.get_depth_scale()
        self.get_logger().info(f'Depth scale: {self.depth_scale:.6f} m/unit')

        # CvBridge
        self.bridge = CvBridge()

        color_topic = self.get_parameter('topics.color_pub').value
        depth_topic = self.get_parameter('topics.depth_pub').value
        depth_colormap_topic = self.get_parameter('topics.depth_colormap_pub').value
        color_info_topic = self.get_parameter('topics.color_info_pub').value
        depth_info_topic = self.get_parameter('topics.depth_info_pub').value

        # Publishers
        self.color_pub = self.create_publisher(Image, color_topic, 10)
        self.depth_pub = self.create_publisher(Image, depth_topic, 10)
        self.depth_colormap_pub = self.create_publisher(Image, depth_colormap_topic, 10)
        self.color_info_pub = self.create_publisher(CameraInfo, color_info_topic, 10)
        self.depth_info_pub = self.create_publisher(CameraInfo, depth_info_topic, 10)

        # Cache for intrinsics (to avoid repeated conversion)
        self._color_intrinsics = None
        self._depth_intrinsics = None
        self._profile = profile

        # main loop timer
        timer_period = 1.0 / self.fps
        self.timer = self.create_timer(timer_period, self.timer_callback)

        self.get_logger().info(
            f'Depth Camera Node started: {self.width}x{self.height} @ {self.fps}fps, '
            f'align={self.align_to_color}'
        )

    def timer_callback(self):
        try:
            frames = self.pipeline.wait_for_frames(timeout_ms=100)
        except Exception:
            self.get_logger().warn('RealSense frame acquisition timeout')
            return

        # Align depth to color if enabled
        if self.align is not None:
            frames = self.align.process(frames)

        now = self.get_clock().now().to_msg()

        # Color image publish
        if self.enable_color:
            color_frame = frames.get_color_frame()
            if color_frame:
                color_image = np.asanyarray(color_frame.get_data())  # (H, W, 3) BGR

                color_msg = self.bridge.cv2_to_imgmsg(color_image, encoding='bgr8')
                color_msg.header = Header(stamp=now, frame_id='camera_color_optical_frame')
                self.color_pub.publish(color_msg)

                # CameraInfo
                if self._color_intrinsics is None:
                    self._color_intrinsics = color_frame.profile.as_video_stream_profile().intrinsics
                self.color_info_pub.publish(
                    self._build_camera_info(self._color_intrinsics, now, 'camera_color_optical_frame')
                )

        # Depth image publish
        if self.enable_depth:
            depth_frame = frames.get_depth_frame()
            if depth_frame:
                depth_image = np.asanyarray(depth_frame.get_data())  # (H, W) uint16, unit = depth_scale m

                # raw depth (uint16, depth_scale meter)
                depth_msg = self.bridge.cv2_to_imgmsg(depth_image, encoding='16UC1')
                depth_msg.header = Header(stamp=now, frame_id='camera_depth_optical_frame')
                self.depth_pub.publish(depth_msg)

                # colormap (for visualization)
                import cv2
                depth_colormap = cv2.applyColorMap(
                    cv2.convertScaleAbs(depth_image, alpha=0.03),
                    cv2.COLORMAP_JET
                )
                colormap_msg = self.bridge.cv2_to_imgmsg(depth_colormap, encoding='bgr8')
                colormap_msg.header = Header(stamp=now, frame_id='camera_depth_optical_frame')
                self.depth_colormap_pub.publish(colormap_msg)

                if self._depth_intrinsics is None:
                    self._depth_intrinsics = depth_frame.profile.as_video_stream_profile().intrinsics
                self.depth_info_pub.publish(
                    self._build_camera_info(self._depth_intrinsics, now, 'camera_depth_optical_frame')
                )

    def _build_camera_info(self, intrinsics, stamp, frame_id: str) -> CameraInfo:
        """RealSense intrinsics -> ROS2 CameraInfo conversion"""
        info = CameraInfo()
        info.header = Header(stamp=stamp, frame_id=frame_id)
        info.width = intrinsics.width
        info.height = intrinsics.height
        info.distortion_model = 'plumb_bob'

        # D: distortion coefficients [k1, k2, p1, p2, k3]
        info.d = list(intrinsics.coeffs)

        # K: camera matrix (row-major, 3x3)
        fx, fy = intrinsics.fx, intrinsics.fy
        cx, cy = intrinsics.ppx, intrinsics.ppy
        info.k = [fx, 0.0, cx,
                  0.0, fy, cy,
                  0.0, 0.0, 1.0]

        # R: rectification matrix (identity for monocular)
        info.r = [1.0, 0.0, 0.0,
                  0.0, 1.0, 0.0,
                  0.0, 0.0, 1.0]

        # P: projection matrix (3x4)
        info.p = [fx, 0.0, cx, 0.0,
                  0.0, fy, cy, 0.0,
                  0.0, 0.0, 1.0, 0.0]
        return info

    def destroy_node(self):
        try:
            self.pipeline.stop()
            self.get_logger().info('RealSense pipeline stopped')
        except Exception:
            pass
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = DepthCameraNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros2_ws/src/perception_camera_cpp/src/depth_camera_node.cpp
#include <chrono>
#include <cstring>
#include <functional>
#include <memory>
#include <string>

#include <cv_bridge/cv_bridge.h>
#include <opencv2/imgproc.hpp>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/camera_info.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <std_msgs/msg/float32.hpp>

#include <librealsense2/rs.hpp>

class DepthCameraNode : public rclcpp::Node {
public:
  DepthCameraNode() : Node("depth_camera_node") {
    width_ = this->declare_parameter<int>("width", 640);
    height_ = this->declare_parameter<int>("height", 480);
    fps_ = this->declare_parameter<int>("fps", 30);
    enable_depth_ = this->declare_parameter<bool>("enable_depth", true);
    enable_color_ = this->declare_parameter<bool>("enable_color", true);
    align_depth_to_color_ = this->declare_parameter<bool>("align_depth_to_color", true);
    depth_scale_publish_ = this->declare_parameter<bool>("depth_scale_publish", true);
    serial_number_ = this->declare_parameter<std::string>("serial_number", "");

    color_pub_ = this->create_publisher<sensor_msgs::msg::Image>("color/image_raw", 10);
    depth_pub_ = this->create_publisher<sensor_msgs::msg::Image>("depth/image_raw", 10);
    color_info_pub_ = this->create_publisher<sensor_msgs::msg::CameraInfo>("color/camera_info", 10);
    depth_info_pub_ = this->create_publisher<sensor_msgs::msg::CameraInfo>("depth/camera_info", 10);
    depth_scale_pub_ = this->create_publisher<std_msgs::msg::Float32>("depth_scale", 10);

    try {
      rs2::config config;
      if (!serial_number_.empty()) {
        config.enable_device(serial_number_);
      }
      if (enable_color_) {
        config.enable_stream(rs2_stream::RS2_STREAM_COLOR, width_, height_, rs2_format::RS2_FORMAT_BGR8, fps_);
      }
      if (enable_depth_) {
        config.enable_stream(rs2_stream::RS2_STREAM_DEPTH, width_, height_, rs2_format::RS2_FORMAT_Z16, fps_);
      }

      profile_ = pipeline_.start(config);

      if (align_depth_to_color_) {
        align_ = std::make_unique<rs2::align>(rs2_stream::RS2_STREAM_COLOR);
      }

      auto depth_sensor = get_depth_sensor(profile_.get_device());
      depth_scale_ = depth_sensor.get_depth_scale();
      RCLCPP_INFO(this->get_logger(), "Depth scale: %.6f m/unit", depth_scale_);

      auto period = std::chrono::duration<double>(1.0 / static_cast<double>(fps_));
      timer_ = this->create_wall_timer(
        std::chrono::duration_cast<std::chrono::milliseconds>(period),
        std::bind(&DepthCameraNode::timer_callback, this));

      RCLCPP_INFO(
        this->get_logger(),
        "Depth Camera Node started: %dx%d @ %dfps, align=%s",
        width_, height_, fps_, align_depth_to_color_ ? "true" : "false");
    } catch (const std::exception & e) {
      RCLCPP_ERROR(this->get_logger(), "Failed to start RealSense pipeline: %s", e.what());
      throw;
    }
  }

  ~DepthCameraNode() override {
    try {
      pipeline_.stop();
      RCLCPP_INFO(this->get_logger(), "RealSense pipeline stopped");
    } catch (...) {
    }
  }

private:
  static rs2::depth_sensor get_depth_sensor(const rs2::device & device) {
    for (const auto & sensor : device.query_sensors()) {
      if (sensor.is<rs2::depth_sensor>()) {
        return sensor.as<rs2::depth_sensor>();
      }
    }
    throw std::runtime_error("No depth sensor found on RealSense device");
  }

  static builtin_interfaces::msg::Time to_builtin_time(const rclcpp::Time & time) {
    builtin_interfaces::msg::Time stamp;
    const int64_t ns = time.nanoseconds();
    stamp.sec = static_cast<int32_t>(ns / 1000000000LL);
    stamp.nanosec = static_cast<uint32_t>(ns % 1000000000LL);
    return stamp;
  }

  sensor_msgs::msg::CameraInfo build_camera_info(const rs2_intrinsics & intrinsics, const builtin_interfaces::msg::Time & stamp, const std::string & frame_id) const {
    sensor_msgs::msg::CameraInfo info;
    info.header.stamp = stamp;
    info.header.frame_id = frame_id;
    info.width = static_cast<uint32_t>(intrinsics.width);
    info.height = static_cast<uint32_t>(intrinsics.height);
    info.distortion_model = "plumb_bob";

    info.d.resize(5);
    for (size_t i = 0; i < 5; ++i) {
      info.d[i] = intrinsics.coeffs[i];
    }

    const double fx = intrinsics.fx;
    const double fy = intrinsics.fy;
    const double cx = intrinsics.ppx;
    const double cy = intrinsics.ppy;

    info.k = {fx, 0.0, cx,
              0.0, fy, cy,
              0.0, 0.0, 1.0};

    info.r = {1.0, 0.0, 0.0,
              0.0, 1.0, 0.0,
              0.0, 0.0, 1.0};

    info.p = {fx, 0.0, cx, 0.0,
              0.0, fy, cy, 0.0,
              0.0, 0.0, 1.0, 0.0};

    return info;
  }

  void timer_callback() {
    rs2::frameset frames;
    try {
      frames = pipeline_.wait_for_frames(100);
    } catch (const std::exception &) {
      RCLCPP_WARN(this->get_logger(), "RealSense frame acquisition timeout");
      return;
    }

    if (align_) {
      frames = align_->process(frames);
    }

    const auto now = to_builtin_time(this->get_clock()->now());

    if (depth_scale_publish_) {
      std_msgs::msg::Float32 msg;
      msg.data = depth_scale_;
      depth_scale_pub_->publish(msg);
    }

    if (enable_color_) {
      const auto color_frame = frames.get_color_frame();
      if (color_frame) {
        const auto w = color_frame.get_width();
        const auto h = color_frame.get_height();
        const auto * color_data = reinterpret_cast<const uint8_t *>(color_frame.get_data());
        cv::Mat color_mat(h, w, CV_8UC3, const_cast<uint8_t *>(color_data), cv::Mat::AUTO_STEP);

        auto color_msg = cv_bridge::CvImage(std_msgs::msg::Header(), "bgr8", color_mat).toImageMsg();
        color_msg->header.stamp = now;
        color_msg->header.frame_id = "camera_color_optical_frame";
        color_pub_->publish(*color_msg);

        if (!color_intrinsics_cached_) {
          auto vsp = color_frame.get_profile().as<rs2::video_stream_profile>();
          color_intrinsics_ = vsp.get_intrinsics();
          color_intrinsics_cached_ = true;
        }
        color_info_pub_->publish(build_camera_info(color_intrinsics_, now, "camera_color_optical_frame"));
      }
    }

    if (enable_depth_) {
      const auto depth_frame = frames.get_depth_frame();
      if (depth_frame) {
        const auto w = depth_frame.get_width();
        const auto h = depth_frame.get_height();
        const auto * depth_data = reinterpret_cast<const uint16_t *>(depth_frame.get_data());

        cv::Mat depth_mat(h, w, CV_16UC1, const_cast<uint16_t *>(depth_data), cv::Mat::AUTO_STEP);
        auto depth_msg = cv_bridge::CvImage(std_msgs::msg::Header(), "16UC1", depth_mat).toImageMsg();
        depth_msg->header.stamp = now;
        depth_msg->header.frame_id = "camera_depth_optical_frame";
        depth_pub_->publish(*depth_msg);

        if (!depth_intrinsics_cached_) {
          auto vsp = depth_frame.get_profile().as<rs2::video_stream_profile>();
          depth_intrinsics_ = vsp.get_intrinsics();
          depth_intrinsics_cached_ = true;
        }
        depth_info_pub_->publish(build_camera_info(depth_intrinsics_, now, "camera_depth_optical_frame"));
      }
    }
  }

  int width_;
  int height_;
  int fps_;
  bool enable_depth_;
  bool enable_color_;
  bool align_depth_to_color_;
  bool depth_scale_publish_;
  std::string serial_number_;

  rs2::pipeline pipeline_;
  rs2::pipeline_profile profile_;
  std::unique_ptr<rs2::align> align_;

  float depth_scale_ {0.001f};
  bool color_intrinsics_cached_ {false};
  bool depth_intrinsics_cached_ {false};
  rs2_intrinsics color_intrinsics_ {};
  rs2_intrinsics depth_intrinsics_ {};

  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr color_pub_;
  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr depth_pub_;
  rclcpp::Publisher<sensor_msgs::msg::CameraInfo>::SharedPtr color_info_pub_;
  rclcpp::Publisher<sensor_msgs::msg::CameraInfo>::SharedPtr depth_info_pub_;
  rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr depth_scale_pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv) {
  rclcpp::init(argc, argv);
  try {
    auto node = std::make_shared<DepthCameraNode>();
    rclcpp::spin(node);
  } catch (const std::exception & e) {
    fprintf(stderr, "Fatal error in depth_camera_node: %s\n", e.what());
  }
  rclcpp::shutdown();
  return 0;
}
